File: calendar2/src/Service/CalendarDatabaseManager.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.Statement;
import java.util.ArrayList;

public class CalendarDatabaseManager 
{   
    Connection connection = null;
    
    public void PreAction()
    {
        try 
        {
            Class.forName("org.sqlite.JDBC");
            connection = DriverManager.getConnection("jdbc:sqlite:BitCalendar.db");
  
            Statement createCalendarEventsStatement = connection.createStatement();
            String sql = "CREATE TABLE IF NOT EXISTS CalendarEvents" +
                   "(UniqueID       INTEGER     PRIMARY KEY     AUTOINCREMENT," +
                   " CreatorID      TEXT, " + 
                   " StartDate      TEXT, " + 
                   " EndDate        TEXT, " + 
                   " Header         TEXT, " + 
                   " Description    TEXT)"; 
            
            createCalendarEventsStatement.executeUpdate(sql);
            createCalendarEventsStatement.close();
            
            Statement createCalendarUsersStatement = connection.createStatement();
            sql = "CREATE TABLE IF NOT EXISTS CalendarUsers" +
                   "(IPAddress TEXT PRIMARY KEY NOT NULL)";

            createCalendarUsersStatement.executeUpdate(sql);
            createCalendarUsersStatement.close();
        } 
        catch (Exception e) 
        {
            System.err.println(e.getClass().getName() + ": " + e.getMessage());
        }
    }  
    public void PostAction()
    {
        try
        {
            connection.close();
        }
        catch (Exception e) 
        {
            System.err.println(e.getClass().getName() + ": " + e.getMessage());
        }        
    }
      
    public boolean CreateCalendarEvent(CalendarEvent calendarEvent)
    {
        System.out.println("Creating calendar event: " + calendarEvent.Header);
        
        boolean result = true;
        
        PreAction();
        
        try
        {
            Statement stmt = connection.createStatement();
            String sql = "INSERT INTO CalendarEvents" + 
                    "(CreatorID,StartDate,EndDate,Header,Description) " +
                    "VALUES (";
            
            if(calendarEvent.CreatorID != null)
                sql += "\"" + calendarEvent.CreatorID   + "\"";
            else
                sql += "\"\"";
            
            if(calendarEvent.StartDate != null)
                sql += ",\"" + calendarEvent.StartDate  + "\"";
            else
                sql += ",\"\"";
            
            if(calendarEvent.EndDate != null)
                sql += ",\"" + calendarEvent.EndDate    + "\"";
            else
                sql += ",\"\"";
            
            if(calendarEvent.Header != null)
                sql += ",\"" + calendarEvent.Header     + "\"";
            else
                sql += ",\"\"";
            
            if(calendarEvent.Description != null)
                sql += ",\"" + calendarEvent.Description + "\"";
            else
                sql += ",\"\"";
            
            sql += ");";
            
            stmt.executeUpdate(sql);
            stmt.close();
      
            System.out.println("Created calendar event: " + calendarEvent.Header);
        }
        catch(Exception e)
        {
            System.out.println("Exception (Create Calendar Event): " + e.getMessage());
        }
         
        PostAction();
        
        return result;
    }
    public boolean EditCalendarEvent(CalendarEvent calendarEvent)
    {
        System.out.println("Editing calendar event: " + calendarEvent.Header);
        
        boolean result = true;
        
        PreAction();
        
        try
        {
            Statement stmt = connection.createStatement();
            String sql = 
                    "UPDATE CalendarEvents SET ";
            
            if (calendarEvent.CreatorID != null)
                sql += "CreatorID = \"" + calendarEvent.CreatorID + "\" , ";
            else
                sql += "CreatorID = \"\" , ";
            
            if (calendarEvent.StartDate != null)
                sql += "StartDate = \"" + calendarEvent.StartDate + "\" , ";
            else
                sql += "StartDate = \"\" , ";
            
            if (calendarEvent.EndDate != null)
                sql += "EndDate = \"" + calendarEvent.EndDate + "\" , ";
            else
                sql += "EndDate = \"\" , ";
            
            if (calendarEvent.Header != null)
                sql += "Header = \"" + calendarEvent.Header + "\" , ";
            else
                sql += "Header = \"\" , ";
            
            if (calendarEvent.Description != null)
                sql += "Description = \"" + calendarEvent.Description + "\" ";
            else
                sql += "Description = \"\" ";
            
            sql += " WHERE UniqueID=" + calendarEvent.UniqueID + ";";
            
            stmt.executeUpdate(sql);
            stmt.close();
            
            System.out.println("Edited calendar event: " + calendarEvent.Header);
        }
        catch(Exception e)
        {
            System.out.println("Exception (Edit Calendar Event): " + e.getMessage());
        }
        
        PostAction();
        
        return result;
    }
    public boolean DeleteCalendarEvent(CalendarEvent calendarEvent)
    {
        System.out.println("Deleting calendar event: " + calendarEvent.Header);
        
        boolean result = true;
        
        PreAction();
        
        try 
        {
            Statement stmt = connection.createStatement();
            String sql = "DELETE from CalendarEvents where UniqueID=" 
                    + calendarEvent.UniqueID + ";";
            stmt.executeUpdate(sql);
            
            System.out.println("Deleted calendar event: " + calendarEvent.Header);
        }
        catch(Exception e)
        {
            System.out.println("Exception (Delete Calendar Event): " + e.getMessage());
        }    
        
        PostAction();
        
        return result;
    }
    public CalendarEvent[] GetCalendarEvents()
    {
        System.out.println("Getting calendar events.");
        
        ArrayList<CalendarEvent> calendarEventsList = new ArrayList<>();

        PreAction();

        try 
        {
            Statement stmt = connection.createStatement();
            ResultSet rs = stmt.executeQuery("SELECT * FROM CalendarEvents;");

            while (rs.next()) 
            {
                int     UniqueID        = rs.getInt("UniqueID");
                String  CreatorID       = rs.getString("CreatorID");
                String  StartDate       = rs.getString("StartDate");
                String  EndDate         = rs.getString("EndDate");
                String  Header          = rs.getString("Header");
                String  Description     = rs.getString("Description");
                
                calendarEventsList.add(new CalendarEvent(   UniqueID,
                                                            CreatorID,
                                                            StartDate,
                                                            EndDate,
                                                            Header,
                                                            Description));
            }   
            rs.close();
            stmt.close();
            
            System.out.println("Got calendar events.");
        }
        catch(Exception e)
        {
            System.out.println("Exception (Get Calendar Events): " + e.getMessage());
        }
 
        PostAction();
        
        return calendarEventsList.toArray(new CalendarEvent[calendarEventsList.size()]);
    }
    
    public boolean CreateUser(CalendarUser calendarUser)
    {
        System.out.println("Creating user: " + calendarUser.IPAddress);
        
        boolean result = true;
        
        PreAction();
        
        try
        {
            Statement stmt = connection.createStatement();
            String sql = "INSERT INTO CalendarUsers" + 
                    "(IPAddress) VALUES (\"" + calendarUser.IPAddress +  "\");";
            stmt.executeUpdate(sql);
            stmt.close();
            System.out.println("Created user: " + calendarUser.IPAddress);     
        }
        catch(Exception e)
        {
            System.out.println("Exception (Create User): " + e.getMessage());
        }
        
        PostAction();
        
        return result;
    }
    public boolean DeleteUser(CalendarUser calendarUser)
    {
        System.out.println("Deleting user: " + calendarUser.IPAddress);
        
        boolean result = true;
        
        PreAction();
        
        try 
        {
            Statement stmt = connection.createStatement();
            String sql = "DELETE from CalendarUsers where IPAddress=\"" 
                    + calendarUser.IPAddress + "\";";
            stmt.executeUpdate(sql);
            System.out.println("Deleted user: " + calendarUser.IPAddress);  
        }
        catch(Exception e)
        {
            System.out.println("Exception (Delete User): " + e.getMessage());
        }    
        
        PostAction();
        
        return result;
    }
    public CalendarUser[] GetUsers()
    {
        System.out.println("Getting users...");
        
        ArrayList<CalendarUser> calendarUsersList = new ArrayList<>();

        PreAction();

        try 
        {
            Statement stmt = connection.createStatement();
            ResultSet rs = stmt.executeQuery("SELECT * FROM CalendarUsers;");

            while (rs.next()) 
            {
                String  IPAddress = rs.getString("IPAddress");
                calendarUsersList.add(new CalendarUser(IPAddress));
            }   
            
            rs.close();
            stmt.close();
            
            System.out.println("Got users.");
        }
        catch(Exception e)
        {
            System.out.println("Exception (Get Users): " + e.getMessage());
        }
 
        PostAction();
        
        return calendarUsersList.toArray(new CalendarUser[calendarUsersList.size()]);
    }
    
    public void ClearCalendarEvents()
    {
        System.out.println("Clearing calendar events table.");
        
        PreAction();
        try 
        {
            Statement stmt = connection.createStatement();
            String sql = "DROP TABLE CalendarEvents";
            stmt.executeUpdate(sql);
            stmt.close();
            System.out.println("Cleared calendar events table.");
        }
        catch(Exception e)
        {
            System.out.println("Exception (Clear Calendar Events): " + e.getMessage());
        }
        
        PostAction();
    }
    public void ClearUsers()
    {
        System.out.println("Clearing calendar users table.");
        
        PreAction();
        
        try 
        {
            Statement stmt = connection.createStatement();
            String sql = "DROP TABLE CalendarUsers";
            stmt.executeUpdate(sql);
            stmt.close();
            System.out.println("Cleared calendar users table.");
        }
        catch(Exception e)
        {
            System.out.println("Exception (Clear Calendar Users): " + e.getMessage());
        }
        
        PostAction();
    }
    public void CreateUsers(CalendarUser[] users)
    {
        for(int i = 0; i < users.length; i++)
            CreateUser(users[i]);
    }
    public void CreateCalendarEvents(CalendarEvent[] calendarEvents)
    {
        for(int i = 0; i < calendarEvents.length; i++)
            CreateCalendarEvent(calendarEvents[i]);
    }
    
}

File: calendar1/src/Service/CalendarDatabaseManager.cs
using System;
using System.Data.SQLite;
using System.Collections.Generic;

namespace BitCalendarService
{
    class CalendarDatabaseManager
    {
        private SQLiteConnection _connection;

        public void PreAction()
        {
            _connection = new SQLiteConnection("Data Source=BitCalendar.db;Version=3;New=True;Compress=True;");

            try
            {
                _connection.Open();
            }
            catch (Exception e)
            {
                Console.WriteLine(e);
                throw;
            }

            var command = _connection.CreateCommand();
            command.CommandText = "CREATE TABLE IF NOT EXISTS CalendarEvents(" +
                                  "UniqueID     INTEGER     PRIMARY KEY     AUTOINCREMENT, " +
                                  "CreatorID    TEXT," +
                                  "StartDate    TEXT," +
                                  "EndDate      TEXT," +
                                  "Header       TEXT," +
                                  "Description  TEXT)";
            command.ExecuteNonQuery();
          
            command = _connection.CreateCommand();
            command.CommandText = "CREATE TABLE IF NOT EXISTS CalendarUsers " +
                                  "(IPAddress    TEXT   PRIMARY KEY     NOT NULL);";
            command.ExecuteNonQuery();
        }
        public void PostAction()
        {
            _connection.Close();
        }

        public bool CreateCalendarEvent(CalendarEvent calendarEvent)
        {
            Console.WriteLine("Creating Calendar Event: " + calendarEvent.Header);

            var result = false;

            PreAction();

            try
            {
                var command = _connection.CreateCommand();
                command.CommandText = "INSERT INTO CalendarEvents" +
                                      "(CreatorID,StartDate,EndDate,Header,Description) " +
                                      "VALUES (";
                if (calendarEvent.CreatorID != null)
                    command.CommandText += "\"" + calendarEvent.CreatorID + "\"";
                else
                    command.CommandText += "\"\"";

                if (calendarEvent.StartDate != null)
                    command.CommandText += ",\"" + calendarEvent.StartDate + "\"" ;
                else
                    command.CommandText += ",\"\"";

                if (calendarEvent.EndDate != null)
                    command.CommandText += ",\"" + calendarEvent.EndDate + "\"";
                else
                    command.CommandText += ",\"\"";

                if (calendarEvent.Header != null)
                    command.CommandText += ",\"" + calendarEvent.Header + "\"";
                else
                    command.CommandText += ",\"\"";

                if (calendarEvent.Description != null)
                    command.CommandText += ",\"" + calendarEvent.Description + "\"";
                else
                    command.CommandText += ",\"\"";

                command.CommandText += ");";

                command.ExecuteNonQuery();
                result = true;
                Console.WriteLine("Created Calendar Event: " + calendarEvent.Header);
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception (Create Calendar Event): " + e.Message);
            }

            PostAction();

            return result;
        }
        public bool EditCalendarEvent(CalendarEvent calendarEvent)
        {
            Console.WriteLine("Editing Calendar Event: " + calendarEvent.Header);

            var result = false;

            PreAction();

            try
            {
                var command = _connection.CreateCommand();
                command.CommandText = "UPDATE CalendarEvents ";

                if (calendarEvent.CreatorID != null)
                    command.CommandText += "SET CreatorID = \"" + calendarEvent.CreatorID + "\" , ";
                else
                    command.CommandText += "SET CreatorID = \"\" , ";

                if (calendarEvent.StartDate != null)
                    command.CommandText += " StartDate = \"" + calendarEvent.StartDate + "\" , ";
                else
                    command.CommandText += " StartDate = \"\" , ";

                if (calendarEvent.EndDate != null)
                    command.CommandText += " EndDate = \"" + calendarEvent.EndDate + "\" , ";
                else
                    command.CommandText += " EndDate = \"\" , ";

                if (calendarEvent.Header != null)
                    command.CommandText += " Header = \"" + calendarEvent.Header + "\" , ";
                else
                    command.CommandText += " Header = \"\" , ";

                if (calendarEvent.Description != null)
                    command.CommandText += " Description = \"" + calendarEvent.Description + "\" ";
                else
                    command.CommandText += " Description = \"\" ";

                command.CommandText += " WHERE UniqueID=" + calendarEvent.UniqueID + ";";

                command.ExecuteNonQuery();
                result = true;
                Console.WriteLine("Edited Calendar Event: " + calendarEvent.Header);
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception (Edit Calendar Event): " + e.Message);
            }

            PostAction();

            return result;
        }
        public bool DeleteCalendarEvent(CalendarEvent calendarEvent)
        {
            Console.WriteLine("Deleting Calendar Event with ID: " + calendarEvent.UniqueID);
            var result = false;

            PreAction();

            try
            {
                var command = _connection.CreateCommand();
                command.CommandText = "DELETE from CalendarEvents where UniqueID="
                                    + calendarEvent.UniqueID + ";";
                result = command.ExecuteNonQuery() >= 0;
                Console.WriteLine("Deleted Calendar Event with ID: " + calendarEvent.UniqueID);
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception (Delete Calendar Event): " + e.Message);
            }

            PostAction();

            return result;
        }
        public CalendarEvent[] GetCalendarEvents()
        {
            Console.WriteLine("Getting Calendar Events... ");

            var calendarEventsList = new List<CalendarEvent>();
                
            PreAction();

            try
            {
                var command = _connection.CreateCommand();
                command.CommandText = "SELECT * FROM CalendarEvents;";

                var reader = command.ExecuteReader();
                while (reader.Read())
                {
                    var uniqueID    = reader.GetInt32(0);
                    var creatorID   = (string)reader["CreatorID"];
                    var startDate   = (string)reader["StartDate"];
                    var endDate     = (string)reader["EndDate"];
                    var header      = (string)reader["Header"];
                    var description = (string)reader["Description"];

                    calendarEventsList.Add(new CalendarEvent() {
                            UniqueID    = uniqueID,
                            CreatorID   = creatorID,
                            StartDate   = startDate,
                            EndDate     = endDate,
                            Header      = header,
                            Description = description });
                }
                Console.WriteLine("Got Calendar Events. ");
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception (Get Calendar Events): " + e.Message);
            }
            
            PostAction();

            return calendarEventsList.ToArray();
        }

        public bool CreateUser(CalendarUser calendarUser)
        {
            Console.WriteLine("Creating Calendar User: " + calendarUser.IPAddress);

            var result = false;

            PreAction();

            try
            {
                var command = _connection.CreateCommand();
                command.CommandText = "INSERT INTO CalendarUsers" +
                    "(IPAddress) VALUES (\"" + calendarUser.IPAddress + "\");";
                command.ExecuteNonQuery();
                result = true;
                Console.WriteLine("Created Calendar User: " + calendarUser.IPAddress);
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception (Create User): " + e.Message);
            }

            PostAction();

            return result;
        }
        public bool DeleteUser(CalendarUser calendarUser)
        {
            Console.WriteLine("Deleting Calendar User: " + calendarUser.IPAddress);

            var result = false;

            PreAction();

            try
            {
                var command = _connection.CreateCommand();
                command.CommandText = "DELETE from CalendarUsers where IPAddress=\""
                                + calendarUser.IPAddress + "\";";
                command.ExecuteNonQuery();
                result = true;
                Console.WriteLine("Deleted Calendar User: " + calendarUser.IPAddress);
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception (Delete User): " + e.Message);
            }

            PostAction();

            return result;
        }
        public CalendarUser[] GetUsers()
        {
            Console.WriteLine("Getting users...");

            var calendarUsersList = new List<CalendarUser>();

            PreAction();

            try
            {
                var command = _connection.CreateCommand();
                command.CommandText = "SELECT * FROM CalendarUsers;";

                var reader = command.ExecuteReader();
                while (reader.Read())
                {
                    var ipAddress = (string)reader["IPAddress"];

                    calendarUsersList.Add(new CalendarUser() { IPAddress = ipAddress });
                }
                Console.WriteLine("Got users.");
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception (Get Users): " + e.Message);
            }

            PostAction();

            return calendarUsersList.ToArray();
        }

        public void ClearCalendarEvents()
        {
            Console.WriteLine("Clearing calendar events table.");

            PreAction();

            try
            {
                var command = _connection.CreateCommand();
                command.CommandText = "DROP TABLE CalendarEvents";
                command.ExecuteNonQuery();
                Console.WriteLine("Cleared calendar events table.");
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception (Clear Calendar Events): " + e.Message);
            }

            PostAction();
        }
        public void ClearUsers()
        {
            Console.WriteLine("Clearing calendar users table.");

            PreAction();

            try
            {
                var command = _connection.CreateCommand();
                command.CommandText = "DROP TABLE CalendarUsers";
                command.ExecuteNonQuery();
                Console.WriteLine("Cleared calendar users table.");
            }
            catch (Exception e)
            {
                Console.WriteLine("Exception (Clear Calendar Users): " + e.Message);
            }

            PostAction();
        }
        public void CreateUsers(CalendarUser[] users)
        {
            foreach (var user in users)
                CreateUser(user);
        }
        public void CreateCalendarEvents(CalendarEvent[] calendarEvents)
        {
            foreach (var calendarEvent in calendarEvents)
                CreateCalendarEvent(calendarEvent);
        }
    }
}
